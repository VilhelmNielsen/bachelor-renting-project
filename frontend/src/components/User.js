import React, { createContext, useContext } from 'react'
import { useQuery, useMutation } from 'react-apollo-hooks'
import graphql from 'graphql-tag'

const MUTATION_SIGNIN = graphql`
  mutation signIn($email: String!, $password: String!) {
    signIn(email: $email, password: $password) {
      id
      name
    }
  }
`

const MUTATION_SIGNOUT = graphql`
  mutation signOut {
    signOut {
      message
    }
  }
`

const MUTATION_SIGNUP = graphql`
  mutation signUp(
    $name: String!
    $email: String!
    $password: String!
    $imageFull: String
    $imagePreview: String
  ) {
    signUp(
      name: $name
      email: $email
      password: $password
      imageFull: $imageFull
      imagePreview: $imagePreview
    ) {
      id
      name
    }
  }
`

const QUERY_CURRENT_USER = graphql`
  query currentUser {
    me {
      id
      name
    }
  }
`

export const userContext = createContext({ user: undefined })

export const useUser = () => useContext(userContext)

const User = {}

function UserProvider(props) {
  const {
    data: { me },
  } = useQuery(QUERY_CURRENT_USER, { errorPolicy: 'ignore' })
  const signUpMutation = useMutation(MUTATION_SIGNUP)
  const signInMutation = useMutation(MUTATION_SIGNIN)
  const signOutMutation = useMutation(MUTATION_SIGNOUT)

  const signIn = async ({ email, password }) => {
    await signInMutation({
      variables: { email, password },
      refetchQueries: [{ query: QUERY_CURRENT_USER }],
    })
  }

  const signOut = async () => {
    await signOutMutation({
      refetchQueries: [{ query: QUERY_CURRENT_USER }],
    })
  }

  const signUp = async ({ email, name, password, imageFull, imagePreview }) => {
    await signUpMutation({
      variables: { email, password, name, imageFull, imagePreview },
      refetchQueries: [{ query: QUERY_CURRENT_USER }],
    })
  }

  return (
    <userContext.Provider
      {...props}
      value={{ user: me, signIn, signOut, signUp }}
    />
  )
}

User.Provider = UserProvider
User.Consumer = userContext.Consumer

export default User
export {
  QUERY_CURRENT_USER,
  MUTATION_SIGNUP,
  MUTATION_SIGNIN,
  MUTATION_SIGNOUT,
}
