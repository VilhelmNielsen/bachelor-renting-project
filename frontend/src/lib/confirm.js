const confirm = window.confirm

export default confirm
