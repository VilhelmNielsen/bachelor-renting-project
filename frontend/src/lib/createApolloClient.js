import ApolloClient from 'apollo-boost'

function createApolloClient() {
  return new ApolloClient({
    uri: process.env.REACT_APP_GRAPHQL_ENDPOINT,
    credentials: 'include',
  })
}

export default createApolloClient
