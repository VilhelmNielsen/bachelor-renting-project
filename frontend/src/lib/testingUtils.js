import casual from 'casual'
import { ApolloClient } from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { MockLink } from 'apollo-link-mock'

casual.seed(123)

export const fakeUser = overrides => ({
  __typename: 'User',
  id: 'abc123',
  name: casual.full_name,
  email: casual.email,
  items: [],
  bookings: [],
  pendingBookings: [],
  ...overrides,
})

export const fakeItem = overrides => ({
  __typename: 'Item',
  id: 'xyz789',
  title: casual.title,
  description: casual.description,
  price: 50000,
  averageRating: null,
  image: {
    full: 'dog_full.jpg',
    preview: 'dog_preview.jpg',
  },
  reviews: [],
  category: [fakeCategory()],
  ...overrides,
})

export const fakeCategory = overrides => ({
  __typename: 'Category',
  id: 'cat345',
  title: casual.title,
  slug: casual.word,
  description: casual.description,
  items: [],
  ...overrides,
})

export function createClient(mocks) {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new MockLink(mocks),
  })
}
