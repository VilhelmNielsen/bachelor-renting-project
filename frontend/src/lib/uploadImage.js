async function uploadImage(event) {
  console.log('Uploading image...')
  const {
    files: [file],
  } = event.target

  const data = new FormData()
  data.append('file', file)
  data.append('upload_preset', 'bachelor-project')

  const res = await fetch(
    'https://api.cloudinary.com/v1_1/vilhelmnielsen/image/upload',
    { method: 'POST', body: data }
  )

  console.log('Upload complete')

  return await res.json()
}

export default uploadImage
