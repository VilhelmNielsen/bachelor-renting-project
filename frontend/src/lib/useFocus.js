import { useState, useRef } from 'react'

function useFocus() {
  const [hasFocus, setHasFocus] = useState(false)
  const timer = useRef()

  const onFocus = event => {
    clearTimeout(timer.current)
    setHasFocus(true)
  }
  const onBlur = event => {
    timer.current = setTimeout(() => setHasFocus(false))
  }

  return [hasFocus, onFocus, onBlur]
}

export default useFocus
