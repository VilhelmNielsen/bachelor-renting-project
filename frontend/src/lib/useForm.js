import { useState } from 'react'

function useForm(initialValues) {
  const [form, setForm] = useState(initialValues)

  const onChange = event => {
    event.persist()
    const { name, value, type } = event.target
    setForm(prevForm => ({
      ...prevForm,
      [name]: type === 'checkbox' ? event.target.checked : value,
    }))
  }

  const reset = () => setForm(initialValues)

  return [form, onChange, reset]
}

export default useForm
