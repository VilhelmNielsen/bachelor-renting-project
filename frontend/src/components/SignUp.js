/** @jsx jsx */
import { jsx, css } from '@emotion/core'
import { useState } from 'react'
import { Form, Label, Input, Fieldset } from './elements/Form'
import Button from './elements/Button'
import useInput from '../lib/useInput'
import ErrorHandler from './Error'
import { useUser } from './User'

function SignUpForm() {
  const { signUp } = useUser()
  const [name, onChangeName, setName] = useInput('')
  const [email, onChangeEmail, setEmail] = useInput('')
  const [password, onChangePassword, setPassword] = useInput('')
  const [
    confirmPassword,
    onChangeConfirmPassword,
    setConfirmPassword,
  ] = useInput('')
  const [busy, setBusy] = useState(false)
  const [error, setError] = useState(null)
  return (
    <Form
      onSubmit={async event => {
        event.preventDefault()
        setBusy(true)
        setError(null)
        try {
          if (password !== confirmPassword) {
            throw new Error("The two password doesn't match!")
          }
          await signUp({ email, password, name })
          setName('')
          setEmail('')
          setPassword('')
          setConfirmPassword('')
          // TODO: Success message
        } catch (error) {
          setError(error)
        } finally {
          setBusy(false)
        }
      }}
    >
      <Fieldset
        css={css`
          display: grid;
          gap: 1rem;
        `}
        disabled={busy}
      >
        <ErrorHandler error={error} />
        <Label htmlFor="signUpName">
          Name
          <Input
            value={name}
            onChange={onChangeName}
            type="text"
            id="signUpName"
            name="name"
          />
        </Label>
        <Label htmlFor="signUpEmail">
          Email
          <Input
            value={email}
            onChange={onChangeEmail}
            type="email"
            id="signUpEmail"
            name="email"
          />
        </Label>
        <Label htmlFor="signUpPassword">
          Password
          <Input
            value={password}
            onChange={onChangePassword}
            type="password"
            id="signUpPassword"
            name="password"
          />
        </Label>

        <Label htmlFor="signUpConfirmPassword">
          Confirm Password
          <Input
            value={confirmPassword}
            onChange={onChangeConfirmPassword}
            type="password"
            id="signUpConfirmPassword"
            name="confirmPassword"
          />
        </Label>

        <Button>Submit</Button>
      </Fieldset>
    </Form>
  )
}

function SignUp() {
  return (
    <div>
      <SignUpForm />
    </div>
  )
}

export default SignUp
export { SignUpForm }
