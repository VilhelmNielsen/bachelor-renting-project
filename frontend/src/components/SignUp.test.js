import React from 'react'
import { render, fireEvent, wait, waitForElement } from 'react-testing-library'
import SignUp from './SignUp'
import { userContext } from './User'
import { fakeUser } from '../lib/testingUtils'

function setup(signUpMock) {
  return render(
    <userContext.Provider value={{ signUp: signUpMock }}>
      <SignUp />
    </userContext.Provider>
  )
}

describe('<SignUp />', () => {
  it('renders and matches snapshot', () => {
    const signUp = jest.fn()
    const { container } = setup(signUp)
    expect(container.querySelector('form')).toHaveFormValues({
      name: '',
      email: '',
      password: '',
      confirmPassword: '',
    })
    expect(container.querySelector('button').textContent).toMatch(/submit/i)
    expect(container).toMatchSnapshot()
  })

  it('can be submitted with data', async () => {
    const signUp = jest.fn()
    const user = fakeUser({ password: 'hunter1' })
    const { container, getByText, getByLabelText } = setup(signUp)

    const nameInput = getByLabelText(/name/i)
    const emailInput = getByLabelText(/email/i)
    const passwordInput = getByLabelText(/password/i)
    const confirmPasswordInput = getByLabelText(/Confirm password/i)

    // Fields are empty
    expect(container.querySelector('form')).toHaveFormValues({
      name: '',
      email: '',
      password: '',
      confirmPassword: '',
    })

    // Fields are filled
    fireEvent.change(nameInput, { target: { value: user.name } })
    fireEvent.change(emailInput, {
      target: { value: user.email },
    })
    fireEvent.change(passwordInput, {
      target: { value: user.password },
    })
    fireEvent.change(confirmPasswordInput, {
      target: { value: user.password },
    })
    expect(container.querySelector('form')).toHaveFormValues({
      name: user.name,
      email: user.email,
      password: user.password,
      confirmPassword: user.password,
    })

    // Fields are disabled when button is clicked
    fireEvent.click(getByText(/submit/i))
    expect(nameInput).toBeDisabled()
    expect(emailInput).toBeDisabled()
    expect(passwordInput).toBeDisabled()
    expect(confirmPasswordInput).toBeDisabled()

    // signUp function is called with the input values
    await wait()
    expect(signUp).toHaveBeenCalledWith({
      name: user.name,
      email: user.email,
      password: user.password,
      imageFull: null,
      imagePreview: null,
    })
    expect(signUp).toHaveBeenCalledTimes(1)

    // Fields are reset and enabled
    expect(container.querySelector('form')).toHaveFormValues({
      name: '',
      email: '',
      password: '',
      confirmPassword: '',
    })
    expect(nameInput).not.toBeDisabled()
    expect(emailInput).not.toBeDisabled()
    expect(passwordInput).not.toBeDisabled()
    expect(confirmPasswordInput).not.toBeDisabled()
  })

  it('will not submit if the confirmPassword does not match the password', async () => {
    const signUp = jest.fn()
    const user = fakeUser({ password: 'hunter1' })
    const { container, getByText, getByLabelText } = setup(signUp)

    const nameInput = getByLabelText(/name/i)
    const emailInput = getByLabelText(/email/i)
    const passwordInput = getByLabelText(/password/i)
    const confirmPasswordInput = getByLabelText(/Confirm password/i)

    fireEvent.change(nameInput, { target: { value: user.name } })
    fireEvent.change(emailInput, {
      target: { value: user.email },
    })
    fireEvent.change(passwordInput, {
      target: { value: user.password },
    })
    fireEvent.change(confirmPasswordInput, {
      target: { value: 'wrongpassword' },
    })
    expect(container.querySelector('form')).toHaveFormValues({
      name: user.name,
      email: user.email,
      password: user.password,
      confirmPassword: 'wrongpassword',
    })

    fireEvent.submit(container.querySelector('form'))
    await waitForElement(() => getByText(/The two password doesn't match!/i), {
      container,
    })
    expect(signUp).not.toHaveBeenCalled()
  })

  it('render errors', async () => {
    const signUp = jest
      .fn()
      .mockRejectedValue({ message: 'Mocked error message' })
    const { container, getByText, debug } = setup(signUp)
    fireEvent.submit(container.querySelector('form'))
    expect(signUp).toHaveBeenCalled()
    await waitForElement(() => getByText(/Error/i), { container })
    expect(getByText(/Mocked error message/i)).toBeInTheDocument()
  })
})
