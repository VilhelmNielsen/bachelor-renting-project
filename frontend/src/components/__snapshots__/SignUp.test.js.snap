// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> renders and matches snapshot 1`] = `
<div>
  <div>
    <form
      class="css-eje67j-Form"
    >
      <fieldset
        class="css-1pwql6a-SignUpForm"
      >
        <label
          class="css-np4was-Label"
          for="signUpName"
        >
          Name
          <input
            class="css-l68vri-Input"
            id="signUpName"
            name="name"
            type="text"
            value=""
          />
        </label>
        <label
          class="css-np4was-Label"
          for="signUpEmail"
        >
          Email
          <input
            class="css-l68vri-Input"
            id="signUpEmail"
            name="email"
            type="email"
            value=""
          />
        </label>
        <label
          class="css-np4was-Label"
          for="signUpImage"
        >
          Image
          <input
            accept="image/jpeg"
            class="css-l68vri-Input"
            id="signUpImage"
            name="image"
            required=""
            type="file"
          />
        </label>
        <label
          class="css-np4was-Label"
          for="signUpPassword"
        >
          Password
          <input
            class="css-l68vri-Input"
            id="signUpPassword"
            name="password"
            type="password"
            value=""
          />
        </label>
        <label
          class="css-np4was-Label"
          for="signUpConfirmPassword"
        >
          Confirm Password
          <input
            class="css-l68vri-Input"
            id="signUpConfirmPassword"
            name="confirmPassword"
            type="password"
            value=""
          />
        </label>
        <button
          class="css-1ajnxt9-Button"
        >
          Submit
        </button>
      </fieldset>
    </form>
  </div>
</div>
`;
